File: apps/ios/ArgusKit/Tests/ArgusKitTests/TestSupport.swift
import Foundation
@testable import ArgusKit

enum TestSupport {
    static func chunk(
        id: String = UUID().uuidString,
        commandId: String = "cmd-1",
        sessionId: String = "sess-1",
        seq: Int,
        kind: ResultKind,
        delta: String? = nil,
        content: String? = nil,
        meta: [String: JSONValue]? = nil,
        isFinal: Bool = false
    ) -> ResultChunk {
        ResultChunk(
            id: id,
            commandId: commandId,
            sessionId: sessionId,
            seq: seq,
            kind: kind,
            delta: delta,
            content: content,
            meta: meta,
            ts: 1_750_000_000_000 + seq,
            isFinal: isFinal
        )
    }

    static func command(
        id: String = "cmd-1",
        sessionId: String = "sess-1",
        kind: CommandKind = .execute,
        prompt: String? = "do the thing",
        status: CommandStatus = .running,
        createdAt: String = "2026-07-05T10:00:00.000Z",
        attachmentIds: [String]? = nil
    ) -> CommandDTO {
        let attachmentsJSON = attachmentIds.map { ids in
            let entries = ids.map { attachmentId in
                """
                {"id": "\(attachmentId)", "filename": "\(attachmentId).png",
                 "mime": "image/png", "size": 1,
                 "url": "/attachments/\(attachmentId)?t=T",
                 "createdAt": "2026-07-05T10:00:00.000Z"}
                """
            }
            return "[\(entries.joined(separator: ","))]"
        }
        let json = """
        {
          "id": "\(id)",
          "sessionId": "\(sessionId)",
          "kind": "\(kind.rawValue)",
          "prompt": \(prompt.map { "\"\($0)\"" } ?? "null"),
          "status": "\(status.rawValue)",
          "createdAt": "\(createdAt)",
          "completedAt": null,
          "attachments": \(attachmentsJSON ?? "null")
        }
        """
        // Decoding keeps the fixture path honest (CommandDTO has no
        // public memberwise init on purpose — DTOs come off the wire).
        // swiftlint:disable:next force_try
        return try! JSONDecoder().decode(CommandDTO.self, from: Data(json.utf8))
    }

    static func session(
        id: String = "sess-1",
        title: String = "Session",
        cliType: String? = "claude-code",
        projectId: String? = nil,
        status: SessionStatus = .idle,
        unread: Bool = false,
        updatedAt: String = "2026-07-05T10:00:00.000Z",
        archivedAt: String? = nil
    ) -> SessionDTO {
        func literal(_ value: String?) -> String {
            value.map { "\"\($0)\"" } ?? "null"
        }
        let json = """
        {
          "id": "\(id)",
          "userId": "u1",
          "projectId": \(literal(projectId)),
          "cliType": \(literal(cliType)),
          "title": "\(title)",
          "externalId": null,
          "status": "\(status.rawValue)",
          "unread": \(unread),
          "archivedAt": \(literal(archivedAt)),
          "createdAt": "2026-07-05T10:00:00.000Z",
          "updatedAt": "\(updatedAt)"
        }
        """
        // Same posture as `command(...)`: decode rather than construct.
        // swiftlint:disable:next force_try
        return try! JSONDecoder().decode(SessionDTO.self, from: Data(json.utf8))
    }

    /// Fixtures live in packages/shared-types/fixtures — ONE directory
    /// shared with the Android client's `:core` tests, so a single
    /// capture keeps both mirrors honest. Resolved from #filePath the way
    /// the lockstep tests resolve the repo root (this file sits six
    /// components below it), so it works under `swift test` on CI and
    /// locally without a SwiftPM resource bundle.
    static var fixturesDirectory: URL {
        URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent() // TestSupport.swift
            .deletingLastPathComponent() // ArgusKitTests
            .deletingLastPathComponent() // Tests
            .deletingLastPathComponent() // ArgusKit
            .deletingLastPathComponent() // ios
            .deletingLastPathComponent() // apps → repo root
            .appending(path: "packages/shared-types/fixtures")
    }

    static func fixtureURL(_ name: String) -> URL {
        fixturesDirectory.appending(path: "\(name).json")
    }

    /// For `.enabled(if:)` on fixtures that are captured on demand — a
    /// test that needs one skips (visibly) until the capture script has
    /// been run against a server that can produce it.
    static func hasFixture(_ name: String) -> Bool {
        FileManager.default.fileExists(atPath: fixtureURL(name).path)
    }

    static func fixtureData(_ name: String) throws -> Data {
        let url = fixtureURL(name)
        guard FileManager.default.fileExists(atPath: url.path) else {
            throw NSError(
                domain: "TestSupport",
                code: 1,
                userInfo: [
                    NSLocalizedDescriptionKey:
                        "missing fixture \(name).json under \(fixturesDirectory.path)"
                ]
            )
        }
        return try Data(contentsOf: url)
    }

    static func decodeFixture<T: Decodable>(_ name: String, as type: T.Type) throws -> T {
        try JSONDecoder().decode(T.self, from: fixtureData(name))
    }
}

File: apps/android/core/src/test/kotlin/app/argus/core/TestSupport.kt
package app.argus.core

import app.argus.core.model.AttachmentDTO
import app.argus.core.model.CommandDTO
import app.argus.core.model.CommandKind
import app.argus.core.model.CommandStatus
import app.argus.core.model.ResultChunk
import app.argus.core.model.ResultKind
import app.argus.core.model.SessionDTO
import app.argus.core.model.SessionStatus
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import java.io.File
import java.util.UUID

// Shared builders and fixture access for :core tests — the counterpart of
// ArgusKit's TestSupport.swift.

fun testChunk(
    id: String = UUID.randomUUID().toString(),
    commandId: String = "cmd-1",
    sessionId: String? = "sess-1",
    seq: Int,
    kind: ResultKind,
    delta: String? = null,
    content: String? = null,
    meta: JsonObject? = null,
    isFinal: Boolean = false,
): ResultChunk = ResultChunk(
    id = id,
    commandId = commandId,
    sessionId = sessionId,
    seq = seq,
    kind = kind,
    delta = delta,
    content = content,
    meta = meta,
    ts = 1_750_000_000_000L + seq,
    isFinal = isFinal,
)

fun testCommand(
    id: String = "cmd-1",
    sessionId: String = "sess-1",
    kind: CommandKind = CommandKind.EXECUTE,
    prompt: String? = "do the thing",
    status: CommandStatus = CommandStatus.RUNNING,
    createdAt: String = "2026-07-05T10:00:00.000Z",
    attachmentIds: List<String>? = null,
): CommandDTO = CommandDTO(
    id = id,
    sessionId = sessionId,
    kind = kind,
    prompt = prompt,
    status = status,
    createdAt = createdAt,
    completedAt = null,
    attachments = attachmentIds?.map { attachmentId ->
        AttachmentDTO(
            id = attachmentId,
            filename = "$attachmentId.png",
            mime = "image/png",
            size = 1,
            url = "/attachments/$attachmentId?t=T",
            createdAt = "2026-07-05T10:00:00.000Z",
        )
    },
)

fun testSession(
    id: String = "sess-1",
    title: String = "Session",
    cliType: String? = "claude-code",
    projectId: String? = null,
    status: SessionStatus = SessionStatus.IDLE,
    unread: Boolean = false,
    updatedAt: String = "2026-07-05T10:00:00.000Z",
    archivedAt: String? = null,
): SessionDTO = SessionDTO(
    id = id,
    userId = "u1",
    projectId = projectId,
    cliType = cliType,
    title = title,
    externalId = null,
    status = status,
    unread = unread,
    archivedAt = archivedAt,
    createdAt = "2026-07-05T10:00:00.000Z",
    updatedAt = updatedAt,
)

/** Parse a JSON object literal into a [JsonObject] (for chunk `meta`). */
fun metaOf(json: String): JsonObject = ArgusJson.parseToJsonElement(json).jsonObject

/**
 * The repository root. Gradle passes it as a system property
 * (core/build.gradle.kts); the walk-up fallback keeps the tests runnable
 * from an IDE that doesn't.
 */
fun repoRoot(): File {
    System.getProperty("argus.repoRoot")?.let { return File(it) }
    var dir: File? = File(System.getProperty("user.dir")).absoluteFile
    while (dir != null) {
        if (File(dir, "pnpm-workspace.yaml").isFile) return dir
        dir = dir.parentFile
    }
    error("could not locate the repository root (no pnpm-workspace.yaml above ${System.getProperty("user.dir")})")
}

/**
 * The fixture directory shared with ArgusKit's FixtureDecodingTests —
 * sanitized live-server responses written by scripts/capture-client-fixtures.sh.
 */
fun fixturesDir(): File =
    System.getProperty("argus.fixtures")?.let { File(it) }
        ?: File(repoRoot(), "packages/shared-types/fixtures")

/**
 * For fixtures that are captured on demand — a test that needs one
 * skips (visibly, via an Assume) until the capture script has been run
 * against a server that can produce it.
 */
fun hasFixture(name: String): Boolean = File(fixturesDir(), "$name.json").isFile

fun fixtureText(name: String): String {
    val file = File(fixturesDir(), "$name.json")
    check(file.isFile) { "missing fixture $name.json under ${fixturesDir()}" }
    return file.readText()
}

inline fun <reified T> decodeFixture(name: String): T = ArgusJson.decodeFromString<T>(fixtureText(name))
